### PR TITLE
Improve connection drop logs

### LINKED ISSUE #1173: Improve logs for server connection drop
### Describe the bug
The server connections have client connections routed into them, and also are used to send data from server to client.
Server connections can drop due to network intermittent issues, or service maintenance. For now, every server connection drop reports an Error. However, some of these Errors actually have no business impact but worry customers because they are **Error** logs. 

Server connections drop with no business impact when:
1. Client->server side: the server connection has no clients routed to
2. Server->Client side: there are still server connections to the same HubServiceEndpoint available

Also, when the server connection having clients routed to drops, shall we make the log more user-friendly to include the impacted client count?

### PR DESCRIPTION
**Pull request by Y-Sindo**
fix #1173

### CODE REVIEW

File: src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.cs [hunk 1]
```diff
@@ -89,6 +89,7 @@ protected override async Task CleanupClientConnections(string fromInstanceId = n
                     connectionIds = _connectionIds.Where(s => s.Value == fromInstanceId).Select(s => s.Key);
                 }
 
+                Log.StartToCleanupClientConnection(Logger, _connectionIds.Count);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.cs:92]: only log when count > 0?

File: src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.Log.cs [hunk 2]
```diff
@@ -27,6 +27,9 @@ private static class Log
             private static readonly Action<ILogger, Exception> _applicationComplete =
                 LoggerMessage.Define(LogLevel.Debug, new EventId(4, "ApplicationComplete"), "Application task completes.");
 
+            private static readonly Action<ILogger, int, Exception> _startToCleanupClientConnections =
+                LoggerMessage.Define<int>(LogLevel.Information, new EventId(5, "StartToCleanupClientConnections"), "Start to cleanup {clientCount} client connections.");
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.Log.cs:31]: explain that it is caused by some specific server connection drop?

File: src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.cs (hunk 1 above)
Comment (Y-Sindo) [reply src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.cs:92]: if count ==0 , the code will return earlier.
https://github.com/Azure/azure-signalr/blob/5c945df8ba0f357e6a7244ab75084c67aa058318/src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.cs#L82-L92

File: src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.cs [hunk 3]
```diff
@@ -88,8 +90,10 @@ protected override async Task CleanupClientConnections(string fromInstanceId = n
                 {
                     connectionIds = _connectionIds.Where(s => s.Value == fromInstanceId).Select(s => s.Key);
                 }
-
-                Log.StartToCleanupClientConnection(Logger, _connectionIds.Count);
+                if (connectionIds.Count() != 0)
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.cs:93]: `Count()` evaluates the collection. `ToArray` first and then use `Length`

File: src/Microsoft.Azure.SignalR.AspNet/ServerConnections/ServiceConnection.Log.cs [hunk 4]
```diff
@@ -11,6 +11,9 @@ internal partial class ServiceConnection
         private static class Log
         {
             // Category: ServiceConnection
+            private static readonly Action<ILogger, int, Exception> _startToCleanupClientConnections =
+                LoggerMessage.Define<int>(LogLevel.Information, new EventId(1, "StartToCleanupClientConnections"), "Start to cleanup {clientCount} client connections, can be triggered by a service connection drop or app server shutdown.");
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.AspNet/ServerConnections/ServiceConnection.Log.cs:15]: `Closing {clientCount} client connections for server connection {serverConnectionId} (and instanceId {fromInstanceId} when it is not null).`

File: src/Microsoft.Azure.SignalR.AspNet/ServerConnections/ServiceConnection.Log.cs [hunk 5]
```diff
@@ -11,6 +11,9 @@ internal partial class ServiceConnection
         private static class Log
         {
             // Category: ServiceConnection
+            private static readonly Action<ILogger, int, string, string, Exception> _closingClientConnections =
+                LoggerMessage.Define<int, string, string>(LogLevel.Information, new EventId(1, "ClosingClientConnections"), "Closing {clientCount} client connections for server connection {serverConnectionId} (and instanceId {fromInstanceId} when it is not null). This can be triggered by a service connection drop or app server shutdown.");
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.AspNet/ServerConnections/ServiceConnection.Log.cs:15]: .......

File: src/Microsoft.Azure.SignalR.AspNet/ServerConnections/ServiceConnection.Log.cs [hunk 6]
```diff
@@ -11,6 +11,12 @@ internal partial class ServiceConnection
         private static class Log
         {
             // Category: ServiceConnection
+            private static readonly Action<ILogger, int, string, string, Exception> _closingClientConnectionsFromInstance =
+                LoggerMessage.Define<int, string, string>(LogLevel.Information, new EventId(1, "ClosingClientConnectionsFromInstance"), "Closing {ClientCount} client connection(s) for server connection {ServerConnectionId} from Azure SignalR instance {InstanceId}.");
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.AspNet/ServerConnections/ServiceConnection.Log.cs:15]: duplicate eventId 1

File: src/Microsoft.Azure.SignalR.AspNet/ServerConnections/ServiceConnection.cs [hunk 7]
```diff
@@ -134,7 +134,19 @@ protected virtual async Task CleanupConnectionsAsyncCore(string instanceId = nul
                 {
                     connectionIds = _clientConnections.Where(s => s.Value.InstanceId == instanceId).Select(s => s.Key);
                 }
-                await Task.WhenAll(connectionIds.Select(s => PerformDisconnectCore(s, true, false)));
+                var tasks = connectionIds.Select(s => PerformDisconnectCore(s, true, false)).ToArray();
+                if (tasks.Length > 0)
+                {
+                    if (string.IsNullOrEmpty(instanceId))
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.AspNet/ServerConnections/ServiceConnection.cs:140]: maybe move the `if` check to Log?

File: src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.Log.cs [hunk 8]
```diff
@@ -27,6 +27,12 @@ private static class Log
             private static readonly Action<ILogger, Exception> _applicationComplete =
                 LoggerMessage.Define(LogLevel.Debug, new EventId(4, "ApplicationComplete"), "Application task completes.");
 
+            private static readonly Action<ILogger, int, string, string, Exception> _closingClientConnectionsFromInstance =
+                LoggerMessage.Define<int, string, string>(LogLevel.Information, new EventId(5, "ClosingClientConnectionsFromInstance"), "Closing {ClientCount} client connection(s) for server connection {ServerConnectionId} from Azure SignalR instance {InstanceId}.");
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.Log.cs:31]: how about reusing `_closingClientConnections`? After all customer is unable to get anything from {InstanceId}...

File: src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.cs [hunk 9]
```diff
@@ -88,8 +90,10 @@ protected override async Task CleanupClientConnections(string fromInstanceId = n
                 {
                     connectionIds = _connectionIds.Where(s => s.Value == fromInstanceId).Select(s => s.Key);
                 }
+                var tasks = connectionIds.Select(PerformDisconnectAsyncCore).ToArray();
 
-                await Task.WhenAll(connectionIds.Select(PerformDisconnectAsyncCore));
+                Log.ClosingClientConnections(Logger, tasks.Length, ConnectionId);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.cs:95]: move `count>0` here sounds better to me

**Review by vicancy — APPROVED** [review]

**Pull request merged**